Mathlib/Order/SuccPred/Archimedean.lean

```diff
@@ -0,0 +1,258 @@
+/-
+Copyright (c) 2021 Yaël Dillies. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies, Daniel Weber
+-/
+import Mathlib.Order.SuccPred.Basic
+
+/-!
+# Archimedean successor and predecessor
+
+* `IsSuccArchimedean`: `SuccOrder` where `succ` iterated to an element gives all the greater
+  ones.
+* `IsPredArchimedean`: `PredOrder` where `pred` iterated to an element gives all the smaller
+  ones.
+-/
+
+variable {α β : Type*}
+
+open Order Function
+
+/-- A `SuccOrder` is succ-archimedean if one can go from any two comparable elements by iterating
+`succ` -/
+class IsSuccArchimedean (α : Type*) [Preorder α] [SuccOrder α] : Prop where
+  /-- If `a ≤ b` then one can get to `a` from `b` by iterating `succ` -/
+  exists_succ_iterate_of_le {a b : α} (h : a ≤ b) : ∃ n, succ^[n] a = b
+
+/-- A `PredOrder` is pred-archimedean if one can go from any two comparable elements by iterating
+`pred` -/
+class IsPredArchimedean (α : Type*) [Preorder α] [PredOrder α] : Prop where
+  /-- If `a ≤ b` then one can get to `b` from `a` by iterating `pred` -/
+  exists_pred_iterate_of_le {a b : α} (h : a ≤ b) : ∃ n, pred^[n] b = a
+
+export IsSuccArchimedean (exists_succ_iterate_of_le)
+
+export IsPredArchimedean (exists_pred_iterate_of_le)
+
+section Preorder
+
+variable [Preorder α]
+
+section SuccOrder
+
+variable [SuccOrder α] [IsSuccArchimedean α] {a b : α}
+
+instance : IsPredArchimedean αᵒᵈ :=
+  ⟨fun {a b} h => by convert exists_succ_iterate_of_le h.ofDual⟩
+
+theorem LE.le.exists_succ_iterate (h : a ≤ b) : ∃ n, succ^[n] a = b :=
+  exists_succ_iterate_of_le h
+
+theorem exists_succ_iterate_iff_le : (∃ n, succ^[n] a = b) ↔ a ≤ b := by
+  refine ⟨?_, exists_succ_iterate_of_le⟩
+  rintro ⟨n, rfl⟩
+  exact id_le_iterate_of_id_le le_succ n a
+
+/-- Induction principle on a type with a `SuccOrder` for all elements above a given element `m`. -/
+@[elab_as_elim]
+theorem Succ.rec {P : α → Prop} {m : α} (h0 : P m) (h1 : ∀ n, m ≤ n → P n → P (succ n)) ⦃n : α⦄
+    (hmn : m ≤ n) : P n := by
+  obtain ⟨n, rfl⟩ := hmn.exists_succ_iterate; clear hmn
+  induction' n with n ih
+  · exact h0
+  · rw [Function.iterate_succ_apply']
+    exact h1 _ (id_le_iterate_of_id_le le_succ n m) ih
+
+theorem Succ.rec_iff {p : α → Prop} (hsucc : ∀ a, p a ↔ p (succ a)) {a b : α} (h : a ≤ b) :
+    p a ↔ p b := by
+  obtain ⟨n, rfl⟩ := h.exists_succ_iterate
+  exact Iterate.rec (fun b => p a ↔ p b) (fun c hc => hc.trans (hsucc _)) Iff.rfl n
+
+lemma le_total_of_ge {r v₁ v₂ : α} (h₁ : r ≤ v₁) (h₂ : r ≤ v₂) : v₁ ≤ v₂ ∨ v₂ ≤ v₁ := by
+  obtain ⟨n, rfl⟩ := h₁.exists_succ_iterate
+  obtain ⟨m, rfl⟩ := h₂.exists_succ_iterate
+  clear h₁ h₂
+  wlog h : n ≤ m
+  · rw [Or.comm]
+    apply this
+    omega
```

Comment:
What's the goal state here? Nothing in this theorem seems to suggest any complex calculations are necessary.

Reply:
I'll replace this with `exact Nat.le_of_not_ge h`.